cfpb/sbl-frontend
# feat(auth): set post_logout_redirect_uri by env var

This PR allows setting the post logout redirect uri via an env variable. It's a small change that can lead to a better logout experience when paired with a keycloak change: allowing users to log out of keycloak then be prompted to log out of login.gov.

This change will allow for users to be redirected to the login.gov user initiated logout for our app after being logged out of keycloak when they click the log out button when combined with:
- setting the `SBL_LOGOUT_REDIRECT_URL` env var to the relying party (user) initiated logout [login.gov url](https://developers.login.gov/oidc/logout/) with the ur's `post_logout_redirect_uri` query param set to redirect back to our home page 
- adding the same RP initiated logout url to keycloak's `Valid post logout redirect URIs` for `regtech-client` (this setting could be [automated](https://www.keycloak.org/operator/realm-import.html) in a later devops PR?)

If a user isn't logged into login.gov, but they hit the RP-initiated logout url, then they just get redirected back to the homepage without hitting the login.gov logout page.

I think the HMDA frontend could also probably implement something very similar by making adjustments to [the logout url in their keycloak.js file](https://github.com/cfpb/hmda-frontend/blob/master/src/common/api/Keycloak.js#L39C3-L39C52)?

In this implementation for SBL, not supplying the `SBL_LOGOUT_REDIRECT_URL` env var defaults back to the old behavior of redirecting back to the `window.location.origin`. 

I've also added types to the env vars as per [the vite guide](https://vitejs.dev/guide/env-and-mode.html#intellisense-for-typescript).

## Changes

- set `post_logout_redirect_uri` to a new env variable `SBL_LOGOUT_REDIRECT_URL`
- add types for the possible env variables as per vite config

## How to test this PR

1. Does logout still works properly when no env variable is set?
2. Can a user be redirected to a new url on logout when the `SBL_LOGOUT_REDIRECT_URL` is set?
    1. Set the `SBL_LOGOUT_REDIRECT_URL` in your `.env` file with a value like `https://consumerfinance.gov/`
    2. Log into Keycloak
    3. Go to the list of clients in the `regtech` realm on your local Keycloak instance ([direct link](http://localhost:8880/admin/master/console/#/regtech/clients))
    4. Click on `regtech-client` 
![Screenshot 2024-04-30 at 4 05 10 AM](https://github.com/cfpb/sbl-frontend/assets/19983248/367852b9-6635-4fc4-9979-50df469e4957)
    5. Change the url you're redirecting to under `Valid post logout redirect URIs`
    ![Screenshot 2024-04-30 at 5 21 10 AM](https://github.com/cfpb/sbl-frontend/assets/19983248/b561e26b-d33b-4a35-9f74-8de03361d4fe)
    6. Log out and see if you get redirected to the new url
3. Does this work with login.gov?
    1. Message me for instructions or see GIF/Demo

## Screenshots

https://github.com/cfpb/sbl-frontend/assets/19983248/ce4753fb-eeab-4b6d-80c7-1c7c206e0051
